profiler/src/ProfilerEngine/Datadog.Profiler.Native/NetworkProvider.cpp

```diff
@@ -0,0 +1,522 @@
+    if (!MonitorRequest(pInfo, pActivityId))
+    {
+        return;
+    }
+
+    // DNS is the first phase...
+    if (pInfo->Redirect == nullptr)
+    {
+        // ... of the first part of the request
+        pInfo->DnsDuration = timestamp - pInfo->DnsStartTime;
+        pInfo->DnsWait = pInfo->DnsStartTime - pInfo->StartTimestamp;
+    }
+    else
+    {
+        // ... of the redirected part of the request
+        pInfo->Redirect->DnsDuration = timestamp - pInfo->Redirect->DnsStartTime;
+        pInfo->Redirect->DnsWait = pInfo->Redirect->DnsStartTime - pInfo->Redirect->StartTimestamp;
+    }
+
+    // even if the first request succeeds, the redirected one might fail
+    pInfo->DnsResolutionSuccess = success;
+}
+
+void NetworkProvider::OnConnectStart(std::chrono::nanoseconds timestamp, LPCGUID pActivityId)
+{
+    NetworkRequestInfo* pInfo = nullptr;
+    if (!MonitorRequest(pInfo, pActivityId, false))
+    {
+        return;
+    }
+
+    if (pInfo->Redirect == nullptr)
+    {
+        pInfo->SocketConnectStartTime = timestamp;
+    }
+    else
+    {
+        pInfo->Redirect->SocketConnectStartTime = timestamp;
+    }
+}
+
+void NetworkProvider::OnConnectStop(std::chrono::nanoseconds timestamp, LPCGUID pActivityId)
+{
+    NetworkRequestInfo* pInfo = nullptr;
+    if (!MonitorRequest(pInfo, pActivityId, false))
+    {
+        return;
+    }
+
+    if (pInfo->Redirect == nullptr)
+    {
+        pInfo->SocketDuration = timestamp - pInfo->SocketConnectStartTime;
+    }
+    else
+    {
+        pInfo->Redirect->SocketDuration = timestamp - pInfo->Redirect->SocketConnectStartTime;
+    }
+}
+
+void NetworkProvider::OnConnectFailed(std::chrono::nanoseconds timestamp, LPCGUID pActivityId, std::string message)
+{
+    NetworkRequestInfo* pInfo = nullptr;
+    if (!MonitorRequest(pInfo, pActivityId, false))
+    {
+        return;
+    }
+
+    if (pInfo->Redirect == nullptr)
+    {
+        pInfo->SocketDuration = timestamp - pInfo->SocketConnectStartTime;
+    }
+    else
+    {
+        pInfo->Redirect->SocketDuration = timestamp - pInfo->Redirect->SocketConnectStartTime;
+    }
+
+    // TODO: check if we should rely on the RequestFailed event to set the error message
+    pInfo->Error = std::move(message);
+}
+
+void NetworkProvider::OnRequestHeaderStart(std::chrono::nanoseconds timestamp, LPCGUID pActivityId)
+{
+    NetworkRequestInfo* pInfo = nullptr;
+    if (!MonitorRequest(pInfo, pActivityId, false))
+    {
+        return;
+    }
+
+    if (pInfo->Redirect == nullptr)
+    {
+        pInfo->ReqRespStartTime = timestamp;
+    }
+    else
+    {
+        pInfo->Redirect->ReqRespStartTime = timestamp;
+    }
+}
+
+// received in .NET 7+
+void NetworkProvider::OnRequestHeaderStop(std::chrono::nanoseconds timestamp, LPCGUID pActivityId, uint32_t statusCode)
+{
+    NetworkRequestInfo* pInfo = nullptr;
+    if (!MonitorRequest(pInfo, pActivityId, false))
+    {
+        return;
+    }
+
+    if (statusCode == 301)
```

Comment:
what is the meaning 301 ? redirect ?

Reply:
I'm extending the code to support all redirect codes